### PR TITLE
[SYCL][COMPAT] Enabled device_info caching within device_ext

### PR DESCRIPTION
**Pull request by OuadiElfarouki**
This patch introduces a simple memoization mechanism within `device_ext` by caching the `device_info`, which helps avoid redundant/unnecessary queries of the full device infos whenever a single device info _(`max_wg_size` for e.g.)_ is requested.

API Tests in `device_ext_api` have been slightly modified to verify some of the expected new changes to `get_device_info` methods.

### CODE REVIEW

File: sycl/test-e2e/syclcompat/device/device.cpp [hunk 1]
```diff
@@ -159,12 +159,26 @@ void test_device_ext_api() {
...
+    assert(Info.get_max_compute_units() == max_cu);
+    assert(Info.get_max_work_group_size() == max_wg_size);
+    assert(Info.get_global_mem_size() == global_mem_size);
+  }
+  {
+    syclcompat::device_info Info = dev_.get_device_info();
+    assert(Info.get_max_compute_units() == max_cu);
+    assert(Info.get_max_work_group_size() == max_wg_size);
+    assert(Info.get_global_mem_size() == global_mem_size);
+  }
```
Comment (Alcpz) [inline sycl/test-e2e/syclcompat/device/device.cpp:180]: Can we improve the test a little bit to ensure the cached path is actually taken?

File: sycl/include/syclcompat/device.hpp [hunk 2]
```diff
@@ -508,31 +511,39 @@ Use 64 bits as memory_bus_width default value."
...
+    } else {
+      out = *_dev_info;
+    }
   }
 
-  device_info get_device_info() const {
-    device_info prop;
-    get_device_info(prop);
-    return prop;
+  const device_info &get_device_info() const {
```
Comment (Alcpz) [inline sycl/include/syclcompat/device.hpp:542]: Adding the `const` qualifier may be problematic. There could be code that uses our headers where:

```
prop = device.get_device_info();
prop.set_...();
```

It seems counterintuitive to modify a device property, but there may be some cases where this happens.
@joeatodd do you remember if this is the case?

File: sycl/doc/syclcompat/README.md [hunk 3]
```diff
@@ -1034,7 +1034,7 @@ class device_ext : public sycl::device {
   void get_memory_info(size_t &free_memory, size_t &total_memory) const;
 
   void get_device_info(device_info &out) const;
-  device_info get_device_info() const;
```
Comment (Alcpz) [inline sycl/doc/syclcompat/README.md:1037]: Revert this change if you end up removing the const qualifier.

**Review by Alcpz — CHANGES_REQUESTED** [review]

File: sycl/include/syclcompat/device.hpp (hunk 2 above)
Comment (OuadiElfarouki) [reply sycl/include/syclcompat/device.hpp:540]: Hmm but in that case we'll need to drop the `const` qualifier of the `get_device_info()` as well? Or return by value which is not ideal. What do you think @Alcpz ?

Comment (Alcpz) [reply sycl/include/syclcompat/device.hpp:540]: You mean the `const` at the end of the signature? Yes, you are right but it's a different issue. You are modifying the value of `_dev_info` in one of the paths.

Comment (OuadiElfarouki) [reply sycl/include/syclcompat/device.hpp]: Hmm noted. I've switched back to the return-by-value to keep it safe for now.

File: sycl/test-e2e/syclcompat/device/device.cpp (hunk 1 above)
Comment (OuadiElfarouki) [reply sycl/test-e2e/syclcompat/device/device.cpp:180]: Can't seem to find any way to improve it at this level (device_info is returned by value).

**Comment by OuadiElfarouki** [comment]
@intel/llvm-gatekeepers Should be good to go! Thank you.
